### PR TITLE
stream: register newshells with no active interpreters

### LINKED ISSUE #29730: ActorGraphInterpreter enter limbo state when interpreter completes but a new shell registration is pending
we've recently deployed a new version of our service, after half a day we've observed unusual heap size and gc activity.

while analyzing a heap dump from our production environment we saw a lot of ActorCell instances, closer look shows that great portion of them were `ActorGraphInterpreter`s.

Examining these actors internal state (as visible in the heap dump) revealed that all of them were in a limbo state where `activeInterpreters` was empty and `newShells` was not and contained a single shell with three logics: SubSource, identity flow and SubSink.
furthermore, the SubSink's internal state was `CancelScheduledBeforeMaterialization` and SubSource's internal state was `ActorSubscriberMessage.OnComplete`.

since the SubSource object's name was referring to `flamMapPrefix` it was obvious this new shell was materialized by `FlatMapPrefix`.
going over our change log between the two deployments we were able to track a new indirect usage of `flatMapPrefix` via `Flow.lazyFutureFlow` which uses the identity flow when upstream completes without producing any elements.

The SubSource's internal state (onComplete message) suggests that `flatMapPrefix`'s upstream actually completed without producing any elements, the internal state of the SubSink also suggest that it was closed prior to being registered (right after materialization).

it seems that `flatMapPrefix` handled both upstreamFinished and downstream cancellation before the materialized internal flow was registered, then when the actor graph interpreter regained control it saw that the current interpreter completed but somehow missed the newShell (speculating here) which put it in a limbo state where it await messages that will never come but not stopped either.

looking at the code I can see two potential issues in the source code:
1. `flatMapPrefix` doesn't set `keepGoing' to true, so once both its ports are closed, the interpreter considers it closed as well.
2. there seems to be a bug in the condition in `akka.stream.impl.fusing.ActorGraphInterpreter#shortCircuitBatch` which doesn't consider `newShells`:
```
while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty)
```
notice that in case all current interpreters are complete but there is one pending newShell, this condition will prevent the actor from seeing the `Resume` message in the `shortCircuitBuffer` which puts the actor int the limbo state described above.

### PR DESCRIPTION
**Pull request by eyalfa**
fixes #29730 .
akka-stream's actor graph interpreter failed to register new shells when current active interpreters where all completed.
The issue was first seen via `Flow.lazyFutureFlow` but is actually a specific issue of the actor graph interpreter exposed via `Flow.flatMapPrefix' and other operators built on top of it (i.e. `Flow.lazyFutureFlow`).

### CODE REVIEW

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala [hunk 1]
```diff
@@ -52,7 +52,7 @@ class PeerSubjectVerifierSpec extends AnyWordSpec with Matchers {
       override def removeValue(name: String): Unit = throw new UnsupportedOperationException()
       override def getValueNames: Array[String] = throw new UnsupportedOperationException()
       override def getLocalCertificates: Array[Certificate] = throw new UnsupportedOperationException()
-      override def getPeerCertificateChain: Array[javax.security.cert.X509Certificate] =
+      override def getPeerCertificateChain/*: Array[javax.security.cert.X509Certificate]*/ =
```
Comment (johanandren) [inline akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala:55]: I don't think we want these in the PR even if you for some mysterious reason needed them on your machine.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala [hunk 2]
```diff
@@ -584,6 +584,35 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
...
+            println("waiting for closer to be set")
+            while (null == closeSink) Thread.sleep(50)
+            println("closing sink")
+            closeSink()
+            println("sink closed")
+            //closing the sink before returning means that it's higly probably
+            //for the flatMapPrefix stage to receive the downstream cancellation before the actor graph interpreter
+            //gets a chance to complete the new interpreter shell's registration.
+            //this in turn exposes a bug in the actor graph interpreter when all active flows complete
+            //but there are pending new interpreter shells to be registered.
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:603]: great with an explanation in here 👍

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala [hunk 3]
```diff
@@ -584,6 +584,35 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
...
+            Flow[Int].prepend(Source(seq))
+          }(Keep.right)
+          .toMat(Sink.queue(10))(Keep.both)
+          .run
+
+        println("assigning closer")
+        closeSink = () => qOut.cancel()
+
+        println("closer assigned, waiting for completion")
+        //Thread.sleep(10000000)
```
Comment (johanandren) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:613]: drop this

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (eyalfa) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala:55]: before dropping this, do u mind checking if you see this as a warning? if this is the case the it'd be a best practice to clean it up

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala (hunk 3 above)
Comment (eyalfa) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:613]:
```suggestion
```

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 4]
```diff
@@ -754,7 +754,7 @@ import akka.util.OptionVal
   }
 
   private def shortCircuitBatch(): Unit = {
-    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty) shortCircuitBuffer
+    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && (activeInterpreters.nonEmpty || newShells.nonEmpty)) shortCircuitBuffer
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: Did you check if removing the `activeInterpreters.nonEmpty` clause has a similar effect?

Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: not sure I'm following, 
but as far as I understand it: in the common case newShells will be empty and activeInterpreters won't. 
so in most case the short circuit buffer holds events relevant for the active interpreters and it doesn't make any sense to drop it.

Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: I meant that using `while (!shortCircuitBuffer.isEmpty && currentLimit > 0)` might suffice. AFAICS the `&& activeInterpreters.nonEmpty` is an optimization that avoids dispatching internal signals that have been scheduled before all interpreters were shutdown. You could say that this optimization was the reason for the bug here because it also prevented running the loop in the forgotten case you found here.

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (johanandren) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: No warning here, on 1.11.0-7, since 1.11.0-8 is not available in the jabba listing yet.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 4 above)
Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: @johanandren linked me to #19299, TBH I haven't followed the entire discussion but it seems the team had good reasons to come up with these optimizations.
when approaching such a delicate and thoughtful code, my approach is to apply the minimal required modification in order to avoid ill effects.

re. the condition and optimization, I generally suspect that in most invocations of this method activeInterpreters will not be empty and newShells will, so the condition will short-circuit. so I believe both my modifications and your suggested modification will have negligible if any effect on performance.

looking at the second condition in this method, it seems to have been designed for cases not covered by the loop, notice it also fail for the scenario in the issue:
```
if (!shortCircuitBuffer.isEmpty && currentLimit == 0) self ! Resume
```

so it might make sense to simplify both conditions in order to guarantee full coverage.

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (eyalfa) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: hmm, this is actually a scalac thing, not java.

Comment (eyalfa) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: my guess, it's related to scalac option `-Xfatal-warnings` which seems to be introduced in AkkaDisciplinePlugin.scala:77 .
this plugin seems to be dominated by a system property `akka.no.discipline`, do you happen to have it defined in your env? does akka's travis/jenkins jobs define it?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 4 above)
Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: > so it might make sense to simplify both conditions in order to guarantee full coverage.

Yes, I agree, we could at least check if all the tests pass with simplified conditions.

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (johanandren) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: The akka-remote module is among the "looseProjects" in there so should not have the strict flags on AFAICs. What is the specific warning/failure you get here with the return type in place?

Comment (raboof) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: Actually 'loose' nowadays just means it doesn't warn about numeric widening, but it's still strict about other things.

I don't see a warning for this line, and I tried a couple of different JVM/Scala version combinations. What JVM and Scala version are you using? What is the exact (fatal) warning you get?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 4 above)
Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: 👍

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (eyalfa) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: this is the warning I get (caret pointing at `: Array`):

> [warn] /Users/efarago/dev/github/akka/akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala:55:45: class X509Certificate in package cert is deprecated: see corresponding Javadoc for more information.
[warn]       override def getPeerCertificateChain: Array[javax.security.cert.X509Certificate] =
[warn]                                             ^
[warn] one warning found

re. scala version, thought I'm using 2.13 seems like sbt has a mind of its own:

> akka > show akka-remote/Test/scalaVersion
[info] 2.12.11

re java version:

> [IJ]akka > reload
[info] welcome to sbt 1.3.13 (AdoptOpenJDK Java 11.0.8)
[info] loading global plugins from /Users/efarago/.sbt/1.0/plugins
[info] loading settings for project akka-build from plugins.sbt,idea1.sbt ...
[info] loading project definition from /Users/efarago/dev/github/akka/project
[info] loading settings for project akka from build.sbt ...

strange thing, it now compiles with the return type in place. the reason I modified this in the first place was due to compilation failures on my env.

to summarize: this line does introduce a warning, do you want to include the modification that suppresses it or not?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 4 above)
Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: @jrudolph , I've just pushed a simplified version of the condition.
however, I think this can be further simplified by converting the `while` into a `@tailrec` method. this way it'd have to exit points, one when the queue is drained, the other when `currentLimit` is reached (with a side effect of sending `Resume` to `self`). I think it'd also be simpler as the existing two current conditions will actually be handled in the same chain (of conditions).
WDUT?

**Comment by raboof** [comment]
> guys, anyone has any idea why travis fail?

it couldn't download the jvm, restarted it.

**Comment by eyalfa** [comment]
guys?

File: akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala (hunk 1 above)
Comment (raboof) [reply akka-remote/src/test/scala/akka/remote/artery/tcp/ssl/PeerSubjectVerifierSpec.scala]: > to summarize: this line does introduce a warning, do you want to include the modification that suppresses it or not?

since this is just the spec I'm OK with keeping the change

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 4 above)
Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: FlatMapMergeBenchmark doesn't seem to show any regression with this change. As you say it's all a bit delicate, but AFAICS it looks good. I'm not sure a `@tailrec` variation would be much simpler, but might be worth a try?

**Review by raboof — APPROVED** [review]

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala [hunk 5]
```diff
@@ -584,6 +584,34 @@ class FlowFlatMapPrefixSpec extends StreamSpec {
...
         subscriber.expectComplete()
       }
+
+      "complete newShells registration when all active interpreters are done" in assertAllStagesStopped {
+        @volatile var closeSink: () => Unit = null
+
+        val (fNotUsed, qOut) = Source
+          .empty[Int]
+          .flatMapPrefixMat(1) { seq =>
+            println("waiting for closer to be set")
```
Comment (jrudolph) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:594]: Can you either remove the println or convert them to `log.debug`?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala [hunk 6]
```diff
@@ -754,8 +754,7 @@ import akka.util.OptionVal
   }
 
   private def shortCircuitBatch(): Unit = {
-    while (!shortCircuitBuffer.isEmpty && currentLimit > 0 && activeInterpreters.nonEmpty) shortCircuitBuffer
-      .poll() match {
+    while (!shortCircuitBuffer.isEmpty && currentLimit > 0) shortCircuitBuffer.poll() match {
```
Comment (jrudolph) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: :+1: this strictly runs more times than before which is good. If it runs even if `activeInterpreters.isEmpty`, `processEvent` below will just skip event processing because `shell.isTerminated` will be false.

**Review by jrudolph — COMMENTED** [review]
Fix LGTM, the test needs a bit of println cleanup.

When I looked into it, I found a few more places where `newShell` registration is broken or gets stuck in edge-cases. I hope to pick that up at some point.

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala (hunk 5 above)
Comment (jrudolph) [reply akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:594]: Also the other ones below.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 6 above)
Comment (eyalfa) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]:
```suggestion
    @tailrec private def shortCircuitBatch(): Unit = {
    if(shortCircuitBuffer.isEmpty) ()
    else if (currentLimit == 0) {
      self ! Resume
    }
    else {
      shortCircuitBuffer.poll() match {
        case b: BoundaryEvent => processEvent(b)
        case Resume           => finishShellRegistration()
      }
      shortCircuitBatch()
    }
  }
```

@jrudolph , how do u feel about making a `tailrec` out of it?

Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala:757]: Sounds good to me :)

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala (hunk 5 above)
Comment (eyalfa) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowFlatMapPrefixSpec.scala:594]:
```suggestion
            log.debug("waiting for closer to be set")
```

File: akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala (hunk 6 above)
Comment (jrudolph) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/ActorGraphInterpreter.scala]: Ah, that suggestion didn't replace the complete old method but only one line of it. Can you fix that?

**Review by jrudolph — APPROVED** [review]
LGTM

**Comment by jrudolph** [comment]
Thanks a lot, @eyalfa!
